src/providers.rs

```diff
@@ -146,52 +210,88 @@ pub fn do_register_provider(caller: Principal, provider: RegisterProviderArgs) -
+        let mut providers = providers.borrow_mut();
+        if let Some(provider) = providers.get(&provider_id) {
+            if provider.owner != caller {
                 ic_cdk::trap("Not authorized");
+            } else {
+                providers.remove(&provider_id).is_some()
             }
         } else {
             false
         }
     })
 }
 
-pub fn do_update_provider(caller: Principal, update: UpdateProviderArgs) {
-    PROVIDERS.with(|p| {
-        let mut p = p.borrow_mut();
-        match p.get(&update.provider_id) {
+/// Changes provider details. The caller must be the owner of the provider.
+pub fn do_update_provider(caller: Principal, args: UpdateProviderArgs) {
+    PROVIDERS.with(|providers| {
+        let mut providers = providers.borrow_mut();
+        match providers.get(&args.provider_id) {
             Some(mut provider) => {
-                if provider.owner != caller && !is_authorized(&caller, Auth::Manage) {
+                if provider.owner != caller {
                     ic_cdk::trap("Provider owner != caller");
+                } else {
+                    if let Some(hostname) = args.hostname {
+                        validate_hostname(&hostname).unwrap();
+                        provider.hostname = hostname;
+                    }
+                    if let Some(path) = args.credential_path {
+                        validate_credential_path(&path).unwrap();
+                        provider.credential_path = path;
+                    }
+                    if let Some(headers) = args.credential_headers {
+                        validate_credential_headers(&headers).unwrap();
+                        provider.credential_headers = headers;
+                    }
+                    if let Some(cycles_per_call) = args.cycles_per_call {
+                        provider.cycles_per_call = cycles_per_call;
+                    }
+                    if let Some(cycles_per_message_byte) = args.cycles_per_message_byte {
+                        provider.cycles_per_message_byte = cycles_per_message_byte;
+                    }
+                    providers.insert(args.provider_id, provider);
                 }
-                if let Some(hostname) = update.hostname {
-                    validate_hostname(&hostname).unwrap();
-                    provider.hostname = hostname;
-                }
-                if let Some(path) = update.credential_path {
-                    validate_credential_path(&path).unwrap();
-                    provider.credential_path = path;
-                }
-                if let Some(headers) = update.credential_headers {
-                    validate_credential_headers(&headers).unwrap();
-                    provider.credential_headers = headers;
+            }
+            None => ic_cdk::trap("Provider not found"),
+        }
+    });
+}
+
+/// Changes administrative details for a provider. The caller must have the `Auth::Manage` permission.
```

Comment:
It's not immediately clear if/where the permission is verified in the code below.